File: merge-tonic-lib/tonic/hydration/memory_index.py
"""In-memory vector index (cosine on hashed char embeddings; parity with TS memoryIndex)."""

from __future__ import annotations

import math
from typing import Any


def text_to_embedding(text: str, dim: int = 48) -> list[float]:
    lower = " ".join(text.lower().split())
    v = [0.0] * dim
    for ch in lower:
        v[ord(ch) % dim] += 1.0
    norm = math.sqrt(sum(x * x for x in v)) or 1.0
    return [x / norm for x in v]


def _dot(a: list[float], b: list[float]) -> float:
    n = min(len(a), len(b))
    return sum(a[i] * b[i] for i in range(n))


class MemoryVectorIndex:
    def __init__(self) -> None:
        self._store: dict[str, dict[str, Any]] = {}

    def upsert(self, records: list[dict[str, Any]]) -> None:
        for r in records:
            rid = str(r["id"])
            self._store[rid] = r

    def query(self, query_embedding: list[float], top_k: int) -> dict[str, Any]:
        entries = list(self._store.values())
        scored: list[tuple[float, dict[str, Any]]] = []
        for r in entries:
            emb = r.get("embedding")
            if isinstance(emb, list) and len(emb) == len(query_embedding):
                s = _dot(emb, query_embedding)
                scored.append((s, r))
            else:
                scored.append((0.0, r))
        scored.sort(key=lambda x: x[0], reverse=True)
        take = scored[:top_k]
        ids = [str(x[1]["id"]) for x in take]
        documents = [str(x[1].get("document", "")) for x in take]
        metadatas = [dict(x[1].get("metadata") or {}) for x in take]
        distances = [1.0 - x[0] for x in take]
        return {"ids": ids, "documents": documents, "metadatas": metadatas, "distances": distances}

    def delete_ids(self, ids: list[str]) -> None:
        for i in ids:
            self._store.pop(i, None)

File: packages/tonic-coding-hydration/src/vector/memoryIndex.ts
import type { QueryResult, VectorIndex, VectorRecord } from "./types";

/** In-memory stub index (dot product on raw vectors when present; else lexical rank 1). */
export function createMemoryVectorIndex(): VectorIndex {
  const store = new Map<string, VectorRecord>();

  return {
    async upsert(records: VectorRecord[]): Promise<void> {
      for (const r of records) {
        store.set(r.id, r);
      }
    },
    async query(queryEmbedding: number[], topK: number): Promise<QueryResult> {
      const entries = [...store.values()];
      const scored = entries.map((r) => {
        if (r.embedding && r.embedding.length === queryEmbedding.length) {
          let s = 0;
          for (let i = 0; i < queryEmbedding.length; i++) {
            s += queryEmbedding[i]! * r.embedding[i]!;
          }
          return { r, score: s };
        }
        return { r, score: 0 };
      });
      scored.sort((a, b) => b.score - a.score);
      const take = scored.slice(0, topK);
      return {
        ids: take.map((x) => x.r.id),
        documents: take.map((x) => x.r.document),
        metadatas: take.map((x) => x.r.metadata),
        distances: take.map((x) => 1 - x.score),
      };
    },
    async deleteIds(ids: string[]): Promise<void> {
      for (const id of ids) {
        store.delete(id);
      }
    },
  };
}
